frontend/exporter/src/id_table.rs

```diff
@@ -0,0 +1,375 @@
+/// This module provides a notion of table, identifiers and nodes. A
+/// `Node<T>` is a `Arc<T>` bundled with a unique identifier such that
+/// there exists an entry in a table for that identifier.
+///
+/// The type `WithTable<T>` bundles a table with a value of type
+/// `T`. That value of type `T` may hold an arbitrary number of
+/// `Node<_>`s. In the context of a `WithTable<T>`, the type `Node<_>`
+/// serializes and deserializes using a table as a state. In this
+/// case, serializing a `Node<U>` produces only an identifier, without
+/// any data of type `U`. Deserializing a `Node<U>` under a
+/// `WithTable<T>` will recover `U` data from the table held by
+/// `WithTable`.
+///
+/// Serde is not designed for stateful (de)serialization. There is no
+/// way of deriving `serde::de::DeserializeSeed` systematically. This
+/// module thus makes use of global state to achieve serialization and
+/// deserialization. This modules provides an API that hides this
+/// global state.
+use crate::prelude::*;
+use std::{
+    hash::{Hash, Hasher},
+    sync::{atomic::Ordering, Arc, LazyLock, Mutex},
+};
+
+/// Unique IDs in a ID table.
+#[derive_group(Serializers)]
+#[derive(Default, Clone, Debug, JsonSchema, Hash, PartialEq, Eq, PartialOrd, Ord)]
+#[serde(transparent)]
+pub struct Id {
+    id: u32,
+}
+
+/// A session providing fresh IDs for ID table.
+#[derive(Default, Debug)]
+pub struct Session {
+    next_id: Id,
+    table: Table,
+}
+
+impl Session {
+    pub fn table(&self) -> &Table {
+        &self.table
+    }
+}
+
+/// The different types of values one can store in an ID table.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub enum Value {
+    Ty(Arc<TyKind>),
```

Comment:
This will keep every type ever translated in memory. I suggest using `Weak` ref-counted pointers instead, so that when all the "real uses" of a given type go out of scope, the type can be deallocated.
```suggestion
    Ty(Weak<TyKind>),
```

Reply:
Mmh, sure, but here that's the state that holds those `Arc`: we allocate a state only during the translation, then we can drop it, and that would drop the table as well, dropping the types held in the table.
Concerning the state of the deserializer, I take care of emptying things after deserialization.

Also, during deserialization, we construct a table incrementally: we'd need to store `Rc`s somewhere before downgrading them to store them in the table.

Reply:
I guess my usecase caches all the `FullDef`s so never drop types without dropping the whole state. If you don't need to drop intermediate types either we can leave this as-is :+1: